### ISSUE TITLE
Fast updating script module can corrupt formatting string passed to parser

### ISSUE THREAD

**Issue by patrick96** [polybar/polybar#1978]
## Describe the issue
When running a script module that updates its contents extremely fast, the formatting string passed to the parser (`polybar -s`) can get corrupted. In particular the formatting string is missing characters and contains unexpected characters (control characters etc).

This was first discovered by [u/runningnak3d](https://www.reddit.com/user/runningnak3d/) on reddit [here](https://www.reddit.com/r/unixporn/comments/egsdro/oc_polybar_vu_meters/fcfpabk) and [here](https://www.reddit.com/r/Polybar/comments/eh7434/polybar_vu_meters/).

**Was it working before?**
* Did you also experience this bug in an earlier version of polybar?
I also tested this with polybar `3.0.0` and was still able to trigger this bug. This has likely been here from the start.

## To Reproduce
```dosini
[bar/example]
modules-left = script

[module/script]
type = custom/script
exec = /tmp/x.sh
tail = true

; vim:set ft=dosini:
```
`/tmp/x.sh`:
```bash
#!/usr/bin/env bash

bars=(
    "=========================================================="
    "==========================================================="
    "============================================================"
    "============================================================="
)

# bars=(
#     "%{F#0f0}==================================="
#     "%{F#ff0}==================================="
#     "%{F#0ff}==================================="
#     "%{F#f00}==================================="
# )

# bars=(
#     "============"
#     "%{F#0f0}==%{F-}=========="
#     "%{F#0f0}====%{F-}========"
#     "%{F#0f0}====%{F-}%{F#ff0}==%{F-}======"
#     "%{F#0f0}====%{F-}%{F#ff0}====%{F-}===="
#     "%{F#0f0}====%{F-}%{F#ff0}====%{F-}%{F#f00}==%{F-}=="
#     "%{F#0f0}====%{F-}%{F#ff0}====%{F-}%{F#f00}====%{F-}"
# )

# bars=(
#     "%{F#0f0}====%{F-}%{F#0f0}============"
#     "%{F#ff0}====%{F-}%{F#0f0}============"
#     "%{F#f00}====%{F-}%{F#0f0}============"
#     "%{F#0ff}====%{F-}%{F#0f0}============"
#     "%{F#00f}====%{F-}%{F#0f0}============"
#     "%{F#f0f}====%{F-}%{F#0f0}============"
# )

while true; do
    for b in ${bars[*]}; do
        echo "$b"
    done
done
```

Since this is likely a race condition bug, you cannot always reliably reproduce this. Try to use some of the other `bars` arrays. From what I have found the bug is easier to reproduce with larger strings, with strings with more formatting tags (because you also get parser errors) and with strings containting multi-byte characters (because you get cairo UTF8 errors). 

## Polybar Log
The bug is not necessarily obvious if you don't look at the log. I was able to get the following warnings and errors:
```
error: cairo_scaled_font_text_to_glyphs()input string not valid UTF-8
warn: Dropping unmatched character (U+0008)
```

I was also able to get a parser error but I can't find the log message and am not able to reproduce that error again.

The cairo error only happens once and after that no text will be rendered on the bar until it is restarted.

## Environment:
* WM: HerbstluftWM
* Distro: ArchLinux
* Output of `polybar -vvv`:

```
polybar 3.4.0-52-g34e742e0

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: /usr/lib/ccache/bin/c++
Compiler flags: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags: -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now 
```

## Additional context

The bug occurs before formatting string is sent to the parser, because the formatting string (`polybar -s`) is already malformed. The expected sequence of formatting strings would be:

```
%{l}==========================================================%{PR}
%{l}============================================================%{PR}
%{l}=============================================================%{PR}
```

But `polybar -s` gave me things like

```
%{l}d==========================================%{PR}
```

With the last `bars` variable I was even able to get

```
%{l0W F#0f0}============%{PR}
```

This one is particularly interesting. I always thought the issue lies in the script module not correctly reading input, but `%{l}` is generated by the controller. But I have found out that the entire `%{l}` is still generated, it just doesn't show up in the output of `polybar -s` because it is followed by the `\b` backspace control character which makes it look like things were deleted even though they still exist in the generated text. I confirmed this by adding the following patch to polybar:

```patch
diff --git a/src/modules/script.cpp b/src/modules/script.cpp
index 3883eb1..5c6587e 100644
--- a/src/modules/script.cpp
+++ b/src/modules/script.cpp
@@ -36,9 +36,15 @@ namespace modules {
             while (!m_stopping && fd != -1 && m_command->is_running() && !io_util::poll(fd, POLLHUP, 0)) {
               if (!io_util::poll_read(fd, 25)) {
                 continue;
-              } else if ((m_output = m_command->readline()) != m_prev) {
-                m_prev = m_output;
-                broadcast();
+              } else {
+                m_output = m_command->readline();
+                for(char c : m_output) {
+                  assert(!iscntrl(c));
+                }
+                if (m_output != m_prev) {
+                  m_prev = m_output;
+                  broadcast();
+                }
               }
             }
 
@@ -169,9 +175,17 @@ namespace modules {
       return "";
     }
 
+    for(char c : m_output) {
+      assert(!iscntrl(c));
+    }
+
     if (m_label) {
       m_label->reset_tokens();
       m_label->replace_token("%output%", m_output);
+
+      for(char c : m_label->get()) {
+        assert(!iscntrl(c));
+      }
     }
 
     string cnt{to_string(m_counter)};
```

This also showed me that the command output is read correctly (at least it doesn't contain control characters). But the first assertion fails in the `get_output` method. So `m_output` is read from `m_command` fine and at some point in between that and `get_output` gets corrupted. I tried putting a `lock_guard` with a new mutex everywhere `m_output` is used, but I got a deadlock and couldn't figure out why.

@Lomadriel do you mind taking a look at this? Maybe you have some insights here.

**Comment by runningnak3d** [comment]
I would just like to add that adding a couple of microseconds delay between updates prevent this bug from being triggered, so maybe the input buffer can't be flushed fast enough? Although that wouldn't explain %{l} getting thrown in.

**Comment by patrick96** [comment]
> so maybe the input buffer can't be flushed fast enough?

I'm not sure what you mean by that

> Although that wouldn't explain %{l} getting thrown in.

`%{l}` is added by the controller, this is expected. The bug occurs inside the script module itself. The contents of `m_output` get corrupted (though I don't know when)

**Comment by runningnak3d** [comment]
> > so maybe the input buffer can't be flushed fast enough?
> 
> I'm not sure what you mean by that

Sorry, I still haven't had a chance to look at the code, but doesn't the module text get stored somewhere before being rendered? Again, I am just guessing, and that is really no help......

I will have time after new years to help *actually* look instead of just guessing.

**Comment by patrick96** [comment]
> Sorry, I still haven't had a chance to look at the code

Don't worry we're all here to learn ;)

> but doesn't the module text get stored somewhere before being rendered?

It does. But that part works fine (because that part is single threaded). The module text is already corrupted inside the module object. Whenever a module changes, it adds an update event to the controller event queue. When the controller processes that event it fetches the contents of all running modules and stitches them together. It then passes that entire string to the parser which then sends the appropriate instructions to the renderer. The module contents fetched by the builder are already corrupt.

**Comment by Lomadriel** [comment]
`m_output` and `m_prev` aren't protected in the lambda and in `get_output`.
I can fix that easily, but I'll need this [PR](https://github.com/polybar/polybar/pull/1790) before, otherwise I'll just have to re-implement it to fix that issue.

> I tried putting a lock_guard with a new mutex everywhere m_output is used, but I got a deadlock and couldn't figure out why.

I guess you get a deadlock, because `m_buildlock` is locked then you try to lock your mutex and in the other thread you lock your mutex then you try to lock `m_buildlock`. (Not really sure about that since I didn't see your code)

**Comment by runningnak3d** [comment]
@patrick96 Here is a new one that I am sure is related. I wrote an indicator to indicate horizontal or vertical tiling direction. Even if I set the interval to 1 (it crashes flat out if set to 0) the bar freezes. It doesn't crash, it isn't even like what is reproducible with the bars. It freezes and just stops updating. It is like it is stuck in a loop that doesn't use CPU. REALLY WEIRD. 

I will post example code ASAP, but I am almost positive that it is related to this issue.
